### PR TITLE
Implement devices in the protocol

### PR DESCRIPTION
**Pull request by loucass003**
Implement real devices in the protcol and fixed how synthetic trackers where listed
Add support for feeder app trackers in the protocol (needed the devices for that)

### CODE REVIEW

**Review by Louka3000 — APPROVED** [review]
Works fine after I tried to break it by doing weird stuff with the HMD role and feeder app controller and tracker.

File: src/main/java/dev/slimevr/gui/TrackersList.java [hunk 1]
```diff
@@ -435,19 +435,16 @@ public void update() {
...
-					int signal = ((IMUTracker) tracker).signalStrength;
-					if (signal == -1) {
-						signalStrength.setText("N/A");
-					} else {
-						// -40 dBm is excellent, -95 dBm is very poor
-						int percentage = (signal - -95) * (100 - 0) / (-40 - -95) + 0;
-						percentage = Math.max(Math.min(percentage, 100), 0);
-						signalStrength.setText(percentage + "% " + "(" + signal + " dBm" + ")");
-					}
+			if (tracker instanceof TrackerWithWireless) {
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/gui/TrackersList.java:438]:
```suggestion

			if (tracker instanceof TrackerWithWireless) {
```

File: src/main/java/dev/slimevr/gui/TrackersList.java [hunk 2]
```diff
@@ -252,7 +252,7 @@ public void actionPerformed(ActionEvent e) {
 					position = new JLabel("0 0 0"),
 					c(1, row, 2, GridBagConstraints.FIRST_LINE_START)
 				);
-			if (tracker instanceof IMUTracker) {
+			if (tracker instanceof TrackerWithWireless) {
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/gui/TrackersList.java:255]:
```suggestion

			if (tracker instanceof TrackerWithWireless) {
```

File: src/main/java/dev/slimevr/gui/TrackersList.java [hunk 3]
```diff
@@ -237,7 +237,7 @@ public void actionPerformed(ActionEvent e) {
 			if (t.hasPosition())
 				add(new JLabel("Position"), c(1, row, 2, GridBagConstraints.FIRST_LINE_START));
 			add(new JLabel("TPS"), c(3, row, 2, GridBagConstraints.FIRST_LINE_START));
-			if (tracker instanceof IMUTracker) {
+			if (tracker instanceof TrackerWithWireless) {
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/gui/TrackersList.java:240]:
```suggestion

			if (tracker instanceof TrackerWithWireless) {
```

File: src/main/java/dev/slimevr/gui/TrackersList.java [hunk 4]
```diff
@@ -435,19 +435,16 @@ public void update() {
...
-					} else {
-						// -40 dBm is excellent, -95 dBm is very poor
-						int percentage = (signal - -95) * (100 - 0) / (-40 - -95) + 0;
-						percentage = Math.max(Math.min(percentage, 100), 0);
-						signalStrength.setText(percentage + "% " + "(" + signal + " dBm" + ")");
-					}
+			if (tracker instanceof TrackerWithWireless) {
+				ping.setText(String.valueOf(((TrackerWithWireless) tracker).getPing()));
+				int signal = ((TrackerWithWireless) tracker).getSignalStrength();
+				if (signal == -1) {
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/gui/TrackersList.java:441]:
```suggestion

				if (signal == -1) {
```

File: src/main/java/dev/slimevr/gui/TrackersList.java [hunk 5]
```diff
@@ -435,19 +435,16 @@ public void update() {
...
+					signalStrength.setText("N/A");
+				} else {
+					// -40 dBm is excellent, -95 dBm is very poor
+					int percentage = (signal - -95) * (100 - 0) / (-40 - -95) + 0;
+					percentage = Math.max(Math.min(percentage, 100), 0);
+					signalStrength.setText(percentage + "% " + "(" + signal + " dBm" + ")");
 				}
 			}
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/gui/TrackersList.java:449]:
```suggestion
			}

```

File: src/main/java/dev/slimevr/platform/windows/WindowsNamedPipeBridge.java [hunk 6]
```diff
@@ -105,13 +106,25 @@ public void changeShareSettings(TrackerRole role, boolean share) {
...
+
 		VRTracker tracker = new VRTracker(
 			trackerAdded.getTrackerId(),
 			trackerAdded.getTrackerSerial(),
 			trackerAdded.getTrackerName(),
 			true,
-			true
+			true,
+			device
 		);
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/platform/windows/WindowsNamedPipeBridge.java:125]:
```suggestion
		);

```

File: src/main/java/dev/slimevr/protocol/DataFeedBuilder.java [hunk 7]
```diff
@@ -172,28 +175,36 @@ public static int createDeviceData(
...
+		HardwareStatus.startHardwareStatus(fbb);
+		HardwareStatus.addErrorStatus(fbb, tracker.getStatus().id);
+
+		if (tracker instanceof TrackerWithTPS)
+			HardwareStatus.addTps(fbb, (int) ((TrackerWithTPS) tracker).getTPS());
+		if (tracker instanceof TrackerWithBattery) {
+			TrackerWithBattery twb = (TrackerWithBattery) tracker;
+			HardwareStatus.addBatteryVoltage(fbb, twb.getBatteryVoltage());
+			HardwareStatus.addBatteryPctEstimate(fbb, (int) twb.getBatteryLevel());
+		}
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/protocol/DataFeedBuilder.java:200]:
```suggestion
		}

```

File: src/main/java/dev/slimevr/vr/Device.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	private String customName;
+	private String firmwareVersion;
+	private String manufacturer;
+
+	private final FastList<Tracker> trackers = new FastList<>();
+
+	public Device() {
+		this.id = nextLocalDeviceId.incrementAndGet();
+	}
+
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/Device.java:19]:
```suggestion
```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 9]
```diff
@@ -222,7 +223,7 @@ private void setUpSensor(UDPDevice connection, int trackerId, int sensorType, in
 					+ " status: "
 					+ sensorStatus
 			);
-		IMUTracker imu = connection.sensors.get(trackerId);
+		IMUTracker imu = connection.getTracker(trackerId);
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:226]:
```suggestion

		IMUTracker imu = connection.getTracker(trackerId);

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 10]
```diff
@@ -233,10 +234,10 @@ private void setUpSensor(UDPDevice connection, int trackerId, int sensorType, in
 				this,
 				Main.vrServer
 			);
-			connection.sensors.put(trackerId, imu);
 			ReferenceAdjustedTracker<IMUTracker> adjustedTracker = new ReferenceAdjustedTracker<>(
 				imu
 			);
+			connection.getTrackers().add(imu);
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:240]:
```suggestion

			connection.getTrackers().add(imu);
```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 11]
```diff
@@ -314,7 +315,7 @@ public void run() {
 							parser.write(bb, conn, new UDPPacket1Heartbeat());
 							socket.send(new DatagramPacket(rcvBuffer, bb.position(), conn.address));
 							if (conn.lastPacket + 1000 < System.currentTimeMillis()) {
-								Iterator<IMUTracker> iterator = conn.sensors.values().iterator();
+								Iterator<IMUTracker> iterator = conn.getTrackers().iterator();
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:318]:
```suggestion
								Iterator<IMUTracker> iterator = conn.getTrackers().iterator();

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 12]
```diff
@@ -326,7 +327,7 @@ public void run() {
 								}
 							} else {
 								conn.timedOut = false;
-								Iterator<IMUTracker> iterator = conn.sensors.values().iterator();
+								Iterator<IMUTracker> iterator = conn.getTrackers().iterator();
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:330]:
```suggestion
								Iterator<IMUTracker> iterator = conn.getTrackers().iterator();

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 13]
```diff
@@ -385,7 +386,7 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 				UDPPacket1Rotation rotationPacket = (UDPPacket1Rotation) packet;
 				buf.set(rotationPacket.rotation);
 				offset.mult(buf, buf);
-				tracker = connection.sensors.get(rotationPacket.getSensorId());
+				tracker = connection.getTracker(rotationPacket.getSensorId());
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:389]:
```suggestion
				tracker = connection.getTracker(rotationPacket.getSensorId());

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 14]
```diff
@@ -395,7 +396,7 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 				if (connection == null)
 					break;
 				UDPPacket17RotationData rotationData = (UDPPacket17RotationData) packet;
-				tracker = connection.sensors.get(rotationData.getSensorId());
+				tracker = connection.getTracker(rotationData.getSensorId());
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:399]:
```suggestion
				tracker = connection.getTracker(rotationData.getSensorId());

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 15]
```diff
@@ -418,7 +419,7 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 				if (connection == null)
 					break;
 				UDPPacket18MagnetometerAccuracy magAccuracy = (UDPPacket18MagnetometerAccuracy) packet;
-				tracker = connection.sensors.get(magAccuracy.getSensorId());
+				tracker = connection.getTracker(magAccuracy.getSensorId());
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:422]:
```suggestion
				tracker = connection.getTracker(magAccuracy.getSensorId());

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 16]
```diff
@@ -439,9 +440,13 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
...
-					for (IMUTracker imuTracker : connection.sensors.values()) {
-						imuTracker.ping = (int) (System.currentTimeMillis()
-							- connection.lastPingPacketTime) / 2;
+					for (IMUTracker imuTracker : connection.getTrackers()) {
+						imuTracker
+							.setPing(
+								(int) (System.currentTimeMillis()
+									- connection.lastPingPacketTime) / 2
+							);
+						;
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:449]:
```suggestion

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 17]
```diff
@@ -478,7 +482,7 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 				if (connection == null)
 					break;
 				UDPPacket13Tap tap = (UDPPacket13Tap) packet;
-				tracker = connection.sensors.get(tap.getSensorId());
+				tracker = connection.getTracker(tap.getSensorId());
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:485]:
```suggestion
				tracker = connection.getTracker(tap.getSensorId());

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 18]
```diff
@@ -500,7 +504,7 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 					);
 				if (connection == null)
 					break;
-				tracker = connection.sensors.get(error.getSensorId());
+				tracker = connection.getTracker(error.getSensorId());
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:507]:
```suggestion

				tracker = connection.getTracker(error.getSensorId());

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 19]
```diff
@@ -464,9 +469,8 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 				if (connection == null)
 					break;
 				UDPPacket12BatteryLevel battery = (UDPPacket12BatteryLevel) packet;
-				if (connection.sensors.size() > 0) {
-					Collection<IMUTracker> trackers = connection.sensors.values();
-					Iterator<IMUTracker> iterator = trackers.iterator();
+				if (connection.getTrackers().size() > 0) {
+					Iterator<IMUTracker> iterator = connection.getTrackers().iterator();
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:473]:
```suggestion

				if (connection.getTrackers().size() > 0) {
					Iterator<IMUTracker> iterator = connection.getTrackers().iterator();

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 20]
```diff
@@ -529,20 +533,19 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
 				if (connection == null)
 					break;
 				UDPPacket19SignalStrength signalStrength = (UDPPacket19SignalStrength) packet;
-				if (connection.sensors.size() > 0) {
-					Collection<IMUTracker> trackers = connection.sensors.values();
-					Iterator<IMUTracker> iterator = trackers.iterator();
+				if (connection.getTrackers().size() > 0) {
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:536]:
```suggestion

				if (connection.getTrackers().size() > 0) {

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 21]
```diff
@@ -529,20 +533,19 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
...
 					}
 				}
 				break;
 			case UDPProtocolParser.PACKET_TEMPERATURE:
 				if (connection == null)
 					break;
 				UDPPacket20Temperature temp = (UDPPacket20Temperature) packet;
-				tracker = connection.sensors.get(temp.getSensorId());
+				tracker = connection.getTracker(temp.getSensorId());
 				if (tracker == null)
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:549]:
```suggestion

				if (tracker == null)
```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]
Just some style changes (why do you guys make your code look so squished?)

**Comment by loucass003** [comment]
> Just some style changes (why do you guys make your code look so squished?)

I havent changed the indent of the original code. If you wat to change that i think i would have to be a separate pr.

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java (hunk 16 above)
Comment (TheDevMinerTV) [reply src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:449]: da hecc

**Comment by TheDevMinerTV** [comment]
Reduced my suggestions to code you've added.

File: src/main/java/dev/slimevr/platform/windows/WindowsNamedPipeBridge.java [hunk 22]
```diff
@@ -105,13 +106,25 @@ public void changeShareSettings(TrackerRole role, boolean share) {
...
+		device.getTrackers().add(tracker);
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/platform/windows/WindowsNamedPipeBridge.java:126]:
```suggestion

		device.getTrackers().add(tracker);
```

File: src/main/java/dev/slimevr/protocol/DataFeedBuilder.java [hunk 23]
```diff
@@ -29,10 +27,15 @@
...
-		IMUTracker tracker = device.sensors.get(0);
+	public static int createHardwareInfo(FlatBufferBuilder fbb, IDevice device) {
+		Tracker tracker = device.getTrackers().get(0).get();
+
+		int nameOffset = device.getFirmwareVersion() != null
+			? fbb.createString(device.getFirmwareVersion())
+			: 0;
 
 		HardwareInfo.startHardwareInfo(fbb);
+		HardwareInfo.addFirmwareVersion(fbb, nameOffset);
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/protocol/DataFeedBuilder.java:38]:
```suggestion
		HardwareInfo.addFirmwareVersion(fbb, nameOffset);

```

File: src/main/java/dev/slimevr/vr/DeviceManager.java [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+		this.server = server;
+	}
+
+	public Device createDevice(String name, String version, String manufacturer) {
+		Device device = new Device();
+
+		device.setCustomName(name);
+		device.setFirmwareVersion(version);
+		device.setManufacturer(manufacturer);
+
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/DeviceManager.java:27]:
```suggestion
```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java (hunk 20 above)
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:536]:
```suggestion

				if (connection.getTrackers().size() > 0) {
```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]

File: src/main/java/dev/slimevr/protocol/DataFeedBuilder.java [hunk 25]
```diff
@@ -172,28 +176,37 @@ public static int createDeviceData(
...
-				tracker.temperature,
-				tracker.getBatteryVoltage(),
-				(int) tracker.getBatteryLevel(),
-				-1
-			);
+		HardwareStatus.startHardwareStatus(fbb);
+		HardwareStatus.addErrorStatus(fbb, tracker.getStatus().id);
+
+		if (tracker instanceof TrackerWithTPS)
+			HardwareStatus.addTps(fbb, (int) ((TrackerWithTPS) tracker).getTPS());
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/protocol/DataFeedBuilder.java:196]:
```suggestion
			HardwareStatus.addTps(fbb, (int) ((TrackerWithTPS) tracker).getTPS());

```

File: src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java [hunk 26]
```diff
@@ -439,9 +441,12 @@ protected void processPacket(DatagramPacket received, UDPPacket packet, UDPDevic
...
 				if (connection.lastPingPacketId == ping.pingId) {
-					for (IMUTracker imuTracker : connection.sensors.values()) {
-						imuTracker.ping = (int) (System.currentTimeMillis()
-							- connection.lastPingPacketTime) / 2;
+					for (IMUTracker imuTracker : connection.getTrackers()) {
+						imuTracker
+							.setPing(
+								(int) (System.currentTimeMillis()
+									- connection.lastPingPacketTime) / 2
+							);
```
Comment (TheDevMinerTV) [inline src/main/java/dev/slimevr/vr/trackers/udp/TrackersUDPServer.java:449]:
```suggestion
							);

```

**Review by TheDevMinerTV — CHANGES_REQUESTED** [review]
minor nitpicking, after this I'm done :D

**Review by TheDevMinerTV — APPROVED** [review]

**Review by Eirenliel — APPROVED** [review]

**Pull request merged**